pallets/foreign-investments/src/tests.rs

```diff
@@ -783,20 +804,85 @@ mod investment {
 				Ok(())
 			});
 
-			util::fulfill_last_swap(Action::Investment, AMOUNT / 2);
+			util::fulfill_last_swap(Action::Investment, foreign_to_pool(AMOUNT / 2));
 
 			assert_eq!(
 				ForeignInvestmentInfo::<Runtime>::get(&USER, INVESTMENT_ID),
 				Some(InvestmentInfo {
 					base: BaseInfo::new(FOREIGN_CURR).unwrap(),
-					decrease_swapped_amount: 0,
+					invested_foreign_amount: AMOUNT / 4,
+					decrease_pending_foreign_amount: 0,
+					decrease_swapped_foreign_amount: 0,
 				})
 			);
 
 			assert_eq!(
 				ForeignInvestment::investment(&USER, INVESTMENT_ID),
+				Ok(AMOUNT / 4)
+			);
+			assert_eq!(
+				MockInvestment::investment(&USER, INVESTMENT_ID),
 				Ok(foreign_to_pool(AMOUNT / 4))
 			);
+		});
+	}
+
+	#[test]
+	fn increase_and_fulfill_and_decrease_and_partial_fulfill_and_partial_increase() {
+		new_test_ext().execute_with(|| {
+			util::base_configuration();
+
+			assert_ok!(ForeignInvestment::increase_foreign_investment(
+				&USER,
+				INVESTMENT_ID,
+				AMOUNT,
+				FOREIGN_CURR
+			));
+
+			util::fulfill_last_swap(Action::Investment, AMOUNT);
+
+			assert_ok!(ForeignInvestment::decrease_foreign_investment(
+				&USER,
+				INVESTMENT_ID,
+				AMOUNT,
+				FOREIGN_CURR
+			));
+
+			util::fulfill_last_swap(Action::Investment, foreign_to_pool(3 * AMOUNT / 4));
+
+			MockDecreaseInvestHook::mock_notify_status_change(|_, msg| {
+				assert_eq!(
+					msg,
+					ExecutedForeignDecreaseInvest {
+						amount_decreased: AMOUNT / 2,
+						foreign_currency: FOREIGN_CURR,
+						amount_remaining: AMOUNT / 2,
+					}
+				);
+				Ok(())
+			});
+
+			assert_ok!(ForeignInvestment::increase_foreign_investment(
+				&USER,
+				INVESTMENT_ID,
+				AMOUNT / 2,
+				FOREIGN_CURR
+			));
```

Comment:
There were a bug (fixed in [5c03bd2](https://github.com/centrifuge/centrifuge-chain/pull/1674/commits/5c03bd217d787dbe3ffb22aded66c43992c96cfc with a solution that I do not like at all).

Basically, there is a rare case where increasing a foreign investment can lead, ironically, into a `Decrease` message. It happens when previously a decrease swap was partially fulfilled, and the increase message cancel the pending part of the previous decrease swap. In that case, a message should be sent to solidity. A more visual example:

- Increase 100 and fully swapped.
- Decrease 100.
- Partial swap with 75 (25 pending). So, we have not sent a `Decrease` msg yet.
- Increase again with 25 (or more), canceling the previous pending swap, and sending here a `Decrease` msg.

I suppose this kind of test does not exist previously in IT, right? @wischli 

The current solution is really difficult to read 😞 . I'll think about how I can clean `pre_increase_swap` and `pre_decrease_swap` because both are getting very complex and require a lot of thinking to know what's really happening.

Reply:
IMO, the `ExecutedDecrease` needs to be dispatched in this case. Otherwise, the decreased 100 are never signalled back to Solidity which relies on the state of Centrifuge Chain. This aligns with the V1 state diagram in which the transition was `ActiveSwapIntoForeignAndSwapIntoForeignDone --increase_investment--> ActiveSwapIntoPoolAndSwapIntoForeignDone --dispatch_ExecutedDecrease--> ActiveSwapIntoPool`.

Reply:
Unfortunately, this test does not exist in integration tests. As part of #1700, I removed a few tests which were handling V1 edge cases that are not edge cases anymore. I don't think the above one was included. Do you think, I should add more integration tests? There is much more boilerplate than with FI unit tests.

Reply:
No worries! I have already handled the UTs.

Reply:
> IMO, the ExecutedDecrease needs to be dispatched in this case with foreign amount 100. Otherwise, the decreased 100 are never signalled back to Solidity which relies on the state of Centrifuge Chain

Following the above example, the message will be signaled back to Solidity once it is increased by at least `25`. In that case, for `25`, solidity will receive a message with `75` amount decreased. But, if we wait until getting `100`, that could never happen.